### DISCUSSION TITLE
bun pm cache rm for packages removed with bun remove

### DISCUSSION THREAD

**Discussion by guest271314** [oven-sh/bun#8078] [Ideas]
Given `bun install esbuild`, do stuff with `esbuild`, then `bun remove esbuild` the `@esbuild` folder still exists in `node_modules`. `package.json` is updated, however after a second `bun install` just to see what happens when `esbuild` is no longer in `package.json` `@esbuild` folder is still in `node_modules` after `remove` and subsequent `install`. 

Either an option to `remove`, e.g., `--cache-remove` (referring to package in `node_modules`), or an option to `install`, e.g., `--force-remove`, or `uninstall` command (remove from `node_modules` and `~/.bun/install/cache`) can be added to explicitly remove removed packages.

**Comment by ddnexus** [comment]
I cannot believe the only option would be to delete node_modules and reinstall. What am I missing?

**Reply by guest271314** [reply]
I don't think we have to delete the whole `node_modules` folder. We do have to roll out own solution. Both to handle to `@package_name` case and deleting directories. The suggested approach from the guides https://bun.sh/guides/write-file/unlink

> To synchronously delete a file with Bun, use the unlinkSync function from the [node:fs](https://nodejs.org/api/fs.html#fs_fs_unlink_path_callback) module. (Currently, there is no Bun API for deleting files.)
>
> To remove a file asynchronously, use the unlink function from the [node:fs/promises](https://nodejs.org/api/fs.html#fs_fspromises_unlink_path) module

**Comment by ddnexus** [comment]
> We do have to roll out own solution. 

OK. So ATM the winner is... deleting the node_module dir and reinstalling 😃

**Reply by guest271314** [reply]
Here's a start

`bun run delete-orphan-packages-in-node-modules.js`

```
import { readdir, unlink, rm } from "node:fs/promises";

const dir = "./node_modules";
const packages = await Bun.file("./package.json").json();
const keys = Object.keys(packages?.dependencies || packages);

if (keys.length === 0) {
  for (const file of await readdir(dir)) {
    if (file === ".bin") {
      continue;
    }
    try {
      await unlink(`${dir}/${file}`);
    } catch (e) {
      console.log(e.message,e.path);
      if (e.message === "Is a directory") {
        await rm(e.path, { recursive: true, force: true });
        continue;
      }
    }

  };
}

if (keys.length) {
  for (const file of await readdir(dir)) {
    console.log({ file });
    if (keys.includes(file)) {
      continue;
    } else {
      if (file === ".bin") {
        continue;
      }
      try {
        await rm(file, { recursive: true, force: true });
    } catch (e) {
        console.log(e.message, e.path);
      }
    }
  }
}
```

**Reply by ddnexus** [reply]
Great! Keep me posted when you will add tests.

**Reply by guest271314** [reply]
You can write the tests, and test for leading `@`, too.

**Reply by pw-64** [reply]
> OK. So ATM the winner is... deleting the node_module dir and reinstalling 😃

`npm prune` will clean the `node_modules` of anything not in the `package.json`
bun doesn't yet have parity for this command
